### PR TITLE
refactor: Consistent accept headers for Github HTTP wrapper

### LINKED ISSUE #6686: Refactor GitHub initRepo to use GraphQL instead of REST
Rewrite `github.initRepo()` so that it uses a repository GraphQL query. The main reason is so that we can retrieve the default branch's commit `sha` without an extra query, and that `sha` will be used as a cache lookup key.

I think this example query captures everything we need:

```graphql
{
  repository(owner: "renovatebot", name: "renovate") {
    isFork
    isLocked
    isMirror
    isPrivate
    isArchived
    nameWithOwner
    mergeCommitAllowed
    rebaseMergeAllowed
    squashMergeAllowed
    defaultBranchRef {
      name
    }
  }
}

```

### PR DESCRIPTION
**Pull request by zharinov**
Ref #6686

### CODE REVIEW

File: lib/platform/github/index.ts [hunk 1]
```diff
@@ -239,14 +251,14 @@ export async function initRepo({
         throw new Error(REPOSITORY_FORKED);
       }
     }
-    if (res.body.full_name && res.body.full_name !== repository) {
+    if (repo?.nameWithOwner && repo?.nameWithOwner !== repository) {
```
Comment (rarkins) [inline lib/platform/github/index.ts:254]: Should we really get this far if `repo` is undefined?

File: lib/util/http/github.ts [hunk 2]
```diff
@@ -223,14 +223,18 @@ export class GithubHttp extends Http<GithubHttpOptions, GithubHttpOptions> {
...
-    accept = 'application/vnd.github.merge-info-preview+json'
-  ): Promise<GithubGraphqlResponse<T>> {
+    options: GraphqlOptions = {}
+  ): Promise<T> {
     let result = null;
 
     const path = 'graphql';
 
+    const {
+      acceptHeader: accept = 'application/vnd.github.merge-info-preview+json',
```
Comment (rarkins) [inline lib/util/http/github.ts:235]: This is only necessary for PRs I think. Do we send it with every request? https://developer.github.com/v4/previews/#mergeinfopreview---more-detailed-information-about-a-pull-requests-merge-state

Comment (zharinov) [reply lib/util/http/github.ts:235]: It was default parameter when this function was private, need to see how much snapshots use it

File: lib/platform/github/index.ts (hunk 1 above)
Comment (zharinov) [reply lib/platform/github/index.ts:254]: Mostly, for tests backward compatibility

Comment (rarkins) [reply lib/platform/github/index.ts:254]: Line 235 already uses `repo.fork` so is it possible that all these other `repo?.` checks are unnecessary?

File: lib/util/http/github.ts (hunk 2 above)
Comment (rarkins) [reply lib/util/http/github.ts:235]: I think ideally it should be provide by the calling function and not added by default. It's only used to enable a certain field in Pull Requests

Comment (zharinov) [reply lib/util/http/github.ts:235]: So for the entire Github HTTP, the default one is `application/vnd.github.v3+json`, but for app mode it's `application/vnd.github.merge-info-preview+json, application/vnd.github.v3+json`? And never `application/json` except being set explicitly?

Comment (rarkins) [reply lib/util/http/github.ts:235]: I think that's correct

Comment (zharinov) [reply lib/util/http/github.ts]: Well, that's pretty tricky to make it consistent with current snapshots, so take a look if `accept`-related changes are good

Comment (rarkins) [reply lib/util/http/github.ts]: It seems you can send only one header (the most specific/detailed) but I prefer to send 2 for future compatibility

File: lib/platform/github/index.ts [hunk 3]
```diff
@@ -273,16 +285,16 @@ export async function initRepo({
       }
     }
     // Use default branch as PR target unless later overridden.
-    config.defaultBranch = res.body.default_branch;
+    config.defaultBranch = repo.defaultBranchRef?.name;
```
Comment (rarkins) [inline lib/platform/github/index.ts:288]:
```suggestion
    config.defaultBranch = repo.defaultBranchRef.name;
```

It's going to crash us later anyway if we have no defaultBranch, so let's remove this and cause the crash to happen immediately.

**Review by rarkins — CHANGES_REQUESTED** [review]
Can you split the header refactoring into its own PR for review and merge first?

**Comment by zharinov** [comment]
Closed in favor of 6736

**Comment by zharinov** [comment]
Closed in favor of #6736

**Pull request closed without merging**